contrib/vcl/source/vcl_io_handle.cc

```diff
@@ -0,0 +1,768 @@
+}
+
+void VclIoHandle::updateEvents(uint32_t events) {
+  auto wrk_index = vclWrkIndexOrRegister();
+  VclIoHandle* vcl_handle = this;
+
+  if (wrk_index && is_listener_) {
+    vcl_handle = wrk_listener_.get();
+  }
+
+  struct epoll_event ev;
+  ev.events = EPOLLET;
+
+  if (events & Event::FileReadyType::Read) {
+    ev.events |= EPOLLIN;
+  }
+  if (events & Event::FileReadyType::Write) {
+    ev.events |= EPOLLOUT;
+  }
+  if (events & Event::FileReadyType::Closed) {
+    ev.events |= EPOLLERR | EPOLLHUP;
+  }
+
+  ev.data.u64 = reinterpret_cast<uint64_t>(vcl_handle);
+
+  vppcom_epoll_ctl(vclEpollHandle(wrk_index), EPOLL_CTL_MOD, vcl_handle->sh(), &ev);
+}
+
+void VclIoHandle::initializeFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                                      Event::FileTriggerType, uint32_t events) {
+  VCL_LOG("adding events for sh {:x} fd {} isListener {}", sh_, fd_, isListener());
+
+  auto wrk_index = vclWrkIndexOrRegister();
+  vclInterfaceRegisterEpollEvent(dispatcher);
+
+  VclIoHandle* vcl_handle = this;
+
+  if (is_listener_) {
+    if (wrk_index) {
+      // If this is not the main worker, make sure a worker listener exists
+      if (!wrk_listener_) {
+        vppcom_endpt_t ep;
+        uint8_t addr_buf[sizeof(struct sockaddr_in6)];
+        ep.ip = addr_buf;
+        uint32_t proto;
+
+        if (peekVclSession(sh_, &ep, &proto)) {
+          RELEASE_ASSERT(0, "");
+        }
+
+        auto address = vclEndptToAddress(ep, -1);
+        auto sh = vppcom_session_create(proto, 1);
+        wrk_listener_ = std::make_unique<VclIoHandle>(static_cast<uint32_t>(sh), VclInvalidFd);
+        wrk_listener_->bind(address);
+        uint32_t rv = vppcom_session_listen(sh, 5);
+        if (rv) {
+          VCL_LOG("listen failed sh {:x}", sh);
+          return;
+        }
+        wrk_listener_->setChildWrkListener(this);
+        VCL_LOG("add worker listener sh {:x} wrk_index {} new sh {:x}", sh_, wrk_index, sh);
+      }
+      vcl_handle = wrk_listener_.get();
+    } else if (not_listened_) {
+      // On main worker, no need to create worker listeners
+      vppcom_session_listen(sh_, 5);
+      not_listened_ = false;
+    }
+  }
+
+  struct epoll_event ev;
+  ev.events = EPOLLET;
+
+  if (events & Event::FileReadyType::Read) {
+    ev.events |= EPOLLIN;
+  }
+  if (events & Event::FileReadyType::Write) {
+    ev.events |= EPOLLOUT;
+  }
+  if (events & Event::FileReadyType::Closed) {
+    ev.events |= EPOLLERR | EPOLLHUP;
+  }
+
+  cb_ = cb;
+  ev.data.u64 = reinterpret_cast<uint64_t>(vcl_handle);
+  vppcom_epoll_ctl(vclEpollHandle(wrk_index), EPOLL_CTL_ADD, vcl_handle->sh(), &ev);
+
+  file_event_ = Event::FileEventPtr{new VclEvent(dispatcher, *vcl_handle, cb)};
+}
+
+IoHandlePtr VclIoHandle::duplicate() {
+  auto wrk_index = vclWrkIndexOrRegister();
+  VCL_LOG("duplicate called");
+
+  if (vppcom_session_worker(sh_) == wrk_index) {
+    return std::unique_ptr<VclIoHandle>(this);
```

Comment:
I'm not sure this is a good idea. Most probably the IoHandle is already owned by something and tracked in another unique pointer.

Reply:
Good catch! Removed code. This used to be needed prior to the recent listener refactor whereby listeners are now allocated upfront on main instead of being duplicated by workers, if memory serves.